README.md
# demo-regularisation
**Regularisation and Cross-Validation of Determinants of Egalitarian Democracy: Demonstration for R**

[![Version](https://img.shields.io/badge/version-v1.2.1-blue.svg)](CHANGELOG.md) [![Project Status: Inactive – The project has reached a stable, usable state but is no longer being actively developed; support/maintenance will be provided as time allows.](https://www.repostatus.org/badges/latest/inactive.svg)](STATUS.md) [![GitHub issues](https://img.shields.io/github/issues/bgonzalezbustamante/demo-regularisation.svg)](https://github.com/bgonzalezbustamante/demo-regularisation/issues/) [![GitHub issues-closed](https://img.shields.io/github/issues-closed/bgonzalezbustamante/demo-regularisation.svg)](https://github.com/bgonzalezbustamante/demo-regularisation/issues?q=is%3Aissue+is%3Aclosed) [![DOI](https://zenodo.org/badge/428344428.svg)](https://zenodo.org/badge/latestdoi/428344428) [![License](https://img.shields.io/badge/license-CC--BY--4.0-black)](LICENSE.md) [![R](https://img.shields.io/badge/made%20with-R%20v4.1.0-1f425f.svg)](https://cran.r-project.org/)

## Overview

This repository contains a demonstration for R of regularisation, cross-validation, and shrinkage methods. The application of these techniques is for illustrative purposes only on determinants of egalitarian democracy. The results should not be considered for any interpretation since it is necessary to develop a causal identification strategy and apply a number of controls, adjustments to standard errors, and robustness checks. In addition, it contains a merged, sliced data set from V-Dem and World Bank (*N* = 27,013). This set includes country-year observations from 1789 to 2019 of 202 countries.

Chunks of this code for extraction and handling V-Dem and World Bank data were used in the following article ([download BibTeX file](refs/BIB-World-Development.bib)):

- González-Bustamante, B. (2021). Early Government Responses to COVID-19 in South America. *World Development, 137*, 105180. DOI: [10.1016/j.worlddev.2020.105180](https://doi.org/10.1016/j.worlddev.2020.105180)

Further details and different applications on this [GitHub repository](https://bgonzalezbustamante.github.io/COVID-19-South-America/) and this OSF-Project (DOI: [10.17605/OSF.IO/6FM7X](https://doi.org/10.17605/OSF.IO/6FM7X)).

## Metadata and Preservation

This code is stored with version control on a GitHub repository. Furthermore, a Digital Object Identifier is provided by Zenodo. The structure of the repository is detailed below.

``` r
demo-regularisation
|-- .gitignore
|-- CHANGELOG.md
|-- demo-regularisation.Rproj
|-- CITATION.cff
|-- CODE_OF_CONDUCT.md
|-- LICENSE.md
|-- README.md
|-- STATUS.md
|-- code
    |-- stage_1_data_cleaning.R
    |-- stage_2_regularisation.R
|-- data
    |-- raw
        |-- vdem_wb.csv
    |-- tidy
        |-- vdem_wb.csv
|-- demo
    |-- regularisation_demo.md
    |-- regularisation_demo.Rmd
    |-- regularisation_demo_files
        |-- figure-gfm
            |-- lasso-1.png
            |-- lasso-2.png
            |-- ols-1.png
            |-- ridge-1.png
            |-- ridge-2.png
|-- results
    |-- tables
        |-- table_01.html
        |-- table_01.tex
        |-- table_02.html
        |-- table_02.tex
|-- refs
    |-- BIB-World-Development.bib
```

10 directories and 24 files.

In addition, this README file in Markdown (MD) format provides specific information to ensure the replicability of the code.

## Storage

The GitHub repository has controlled access with Two-Factor Authentication (2FA) with two physical USB security devices (Bastián González-Bustamante, [ORCID iD 0000-0003-1510-6820](https://orcid.org/0000-0003-1510-6820)). Both USB keys issue one-time passwords to generate a cryptographic authentication FIDO2 and U2F.

## Getting Started

### Software

We use R version 4.1.0 (2021-05-18) -- "Camp Pontanezen".

Required R libraries are: "broom", "caret", "coefplot", "glmnet", "tidyverse", "stargazer", and "wbstats".

We recommend that users run replication code and scripts from the root directory using the R project "**demo-regularisation.Rproj**".

### Replication Instructions

Folder "code" contains the R scripts.

Folder "demo" contains a demonstration in RMD and MD formats (**[regularisation_demo.md](demo/regularisation_demo.md)**). On the other hand, "results/tables" includes all tables provided as HTML and TeX files.

These files will be overwritten if you reproduced the steps described below. 

- **Stage 1.** Run R script "**stage_1_data_cleaning.R**" from the "code" folder. This script splits V-Dem data[^1] and merges them with World Bank indicators[^2] on GDP growth and inflation. Then, a significantly smaller and more manageable data set is saved in CSV UTF-8 format (1.57 MB) on this repository.

- **Stage 2.** Run R script "**stage_2_regularisation.R**" from the "code" folder. This script contains the demonstration for R. Alternatively, it is possible to run "**regularisation_demo.Rmd**" from "demo" and the files in "demo/regularisation_demo_files" subfolder will be overwritten.

It is possible to run the code from the second stage onward to check the methods directly. Considering the volume of V-Dem data (182 MB), running the first script takes some time.

### Codebook

The file "**vdem_wb.csv**" in "data/tidy" subfolder is the merged, sliced data set from V-Dem and World Bank (*N* = 27,013). This set contains country-year observations from 1789 to 2019 of 202 countries.[^3]

- **country** (country_name in V-Dem). Country name.

- **year**. Year variable.

- **egal_dem** (v2x_egaldem in V-Dem). The egalitarian democracy index considers freedoms protected across all social groups, resources distributed equally across all social groups, and equal access to power. It also takes into account the level of electoral democracy.

- **corruption** (v2x_corr in V-Dem). The political corruption index measures how pervasive is political corruption and considers measures of six distinct types of corruption from different areas of the political field, distinguishing between executive, legislative, and judicial corruption.

- **military** (v2x_ex_military in V-Dem). The military dimension index measures if the military determines the chief executive's power base based on appointments made through a coup or rebellions and if the military can remove them.

- **free_exp** (v2x_freexp in V-Dem). The freedom of expression index reflects the government's level of respect for press and media freedom, the freedom to discuss political matters in the public sphere, and freedom of academic and cultural expressions.

- **fed_uni** (v2x_feduni in V-Dem). The division of power index reflects if the local and regional governments are elected and the level of independence in the decision-making process.

- **inflation** (FP.CPI.TOTL.ZG in World Bank API). Inflation based on consumer prices (annual percentage).

- **gdp** (NY.GDP.MKTP.KD.ZG in World Bank API). GDP growth (annual percentage).

- **gdp_pc** (NY.GDP.PCAP.KD.ZG in World Bank API). GDP per capita growth (annual percentage).

## License

This R code and merged, sliced data set from V-Dem and World Bank are released under a [Creative Commons Attribution 4.0 International license (CC BY 4.0)](LICENSE.md). This open-access license allows the data to be shared, reused, adapted as long as appropriate acknowledgement is given.

## Contribute

Contributions are entirely welcome. You just need to [open an issue](https://github.com/bgonzalezbustamante/demo-regularisation/issues/new) with your comment or idea.

For more substantial contributions, please fork this repository and make changes. Pull requests are also welcome.

Please read our [code of conduct](CODE_OF_CONDUCT.md) first. Minor contributions will be acknowledged, and significant ones will be considered on our contributor roles taxonomy.

## Citation

González-Bustamante, B. (2022). Regularisation and Cross-Validation of Determinants of Egalitarian Democracy: Demonstration for R (Version 1.2.1 -- White Waterfall) [Computer software]. DOI: [10.5281/zenodo.5708892](https://doi.org/10.5281/zenodo.5708892)

## Author

Bastián González-Bustamante \
bastian.gonzalezbustamante@politics.ox.ac.uk \
[ORCID iD 0000-0003-1510-6820](https://orcid.org/0000-0003-1510-6820) \
https://bgonzalezbustamante.com 

## CRediT - Contributor Roles Taxonomy

Bastián González-Bustamante ([ORCID iD 0000-0003-1510-6820](https://orcid.org/0000-0003-1510-6820)): Conceptualisation, data curation, formal analysis, methodology, project administration, resources, software, validation, and visualisation.

### Latest Revision

[February 12, 2022](CHANGELOG.md).

[^1]: V-Dem [Country–Year/Country–Date] Dataset v10 (Coppedge et al., 2020; DOI: [10.23696/vdemds20](https://doi.org/10.23696/vdemds20)) is downloaded from our OSF-project on COVID-19 in South America (DOI: [10.17605/OSF.IO/6FM7X](https://doi.org/10.17605/OSF.IO/6FM7X)) during the first stage. However, it is important to bear in mind that the current data set is v11.1.
[^2]: Data downloaded during the first stage from the World Bank API.
[^3]: This data set has been updated running the first stage code on February 12, 2022 (v1.2.1 -- White Waterfall). Therefore, there are slight differences with the previous version compiled on November 30, 2021 (v1.1.1 – Autumn Mode). The latter is used in the demonstration and is stored in the “data/raw” subfolder.


CHANGELOG.md
# Changelog
All notable changes to "Regularisation and Cross-Validation of Determinants of Egalitarian Democracy: Demonstration for R" will be documented in this file.

## [Unreleased]

## [Released]

## [1.2.1] - 2022-02-12
### Added
- BibTeX file.
### Changed
- Scripts and folders names standardisation.
- Minor corrections.

## [1.1.1] - 2021-11-30
### Added
- Code of conduct in MD format.
### Changed
- Minor corrections.

## [1.0.1] - 2021-11-18
### Changed
- Minor corrections.

## [1.0.0] - 2021-11-17
### Added
- Repository status badge.
- Citation CFF file.
- Data cleaning code.
- Zenodo DOI.
### Changed
- OLS regularisation code 
- Cross-Validation procedures code.
- Shrinkage methods code 
- Revision of upgrade versioning.
- Minor corrections.

## [0.2.2] - 2021-11-16
### Changed
- Minor corrections.

## [0.2.1] - 2021-11-15
### Added
- GitHub repository.
- README in MD format on GitHub repository.
- Badges in SVG on GitHub repository.
- License CC-BY Attribution 4.0 International on GitHub repository.
- Changelog in MD format on GitHub repository.

## [0.1.1] - 2020-01-20
### Changed
- OLS regularisation code 
- Cross-Validation procedures code.

## [0.1.0] - 2019-12-05
### Added
- OLS regularisation code from OXF-Training repository.
- Cross-Validation procedures code from OXF-Training repository.
- Shrinkage methods code from OXF-Training repository.

code/stage_2_regularisation.R
## +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

#### Script ID ####

## +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

## Script Data Cleaning: V-Dem and World Bank
## R version 4.1.0 (2021-05-18) -- "Camp Pontanezen"
## Date: November 2021

## Bastián González-Bustamante (University of Oxford, UK)
## ORCID iD 0000-0003-1510-6820
## https://bgonzalezbustamante.com

## Regularisation and Cross-Validation: Demonstration for R
## https://github.com/bgonzalezbustamante/demo-regularisation/

## +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

#### Packages and Data ####

## +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

## Clean Environment
rm(list = ls())

## Packages
library(caret)
library(tidyverse)
library(coefplot)
library(stargazer)
library(glmnet)
library(broom)

## Data
vdem_wb <- read.csv("data/raw/vdem_wb.csv", encoding = "UTF-8")

## Descriptives
nrow(vdem_wb)
min(vdem_wb$year)
max(vdem_wb$year)
length(unique(vdem_wb$country))

## Drop NAs
vdem_wb <- na.omit(vdem_wb) 

## +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

#### OLS Regularisation ####

## +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

## Training Set (80%) and Test Set (20%)
set.seed(20211118)
train_index <- createDataPartition(vdem_wb$egal_dem, p = 0.8, list = FALSE)
train_set <- as.data.frame(vdem_wb[train_index,])
test_set <- as.data.frame(vdem_wb[-train_index,])

## Linear Models on the Train Set
slm_train <- lm(egal_dem ~ gdp, data = train_set)
mlm_train <- lm(egal_dem ~ ., data = train_set[-1])

## Coefplot
coefplot(mlm_train, title = "Coefficient Plot Training Set", color = "tomato")

## Review Predicted Observations on the Test Set
slm_train_test <- predict(slm_train, test_set)
mlm_train_test <- predict(mlm_train, test_set)

## Calculate Root Mean Squared Error (RMSE)
rmse_slm <- data.frame(predicted = slm_train_test, actual = test_set$egal_dem, 
                     se = ((slm_train_test - test_set$egal_dem) ^ 2 / 
                             length(slm_train_test)))

sqrt(sum(rmse_slm$se))

rmse_mlm <- data.frame(predicted = mlm_train_test, actual = test_set$egal_dem, 
                     se = ((mlm_train_test - test_set$egal_dem)^2 /
                             length(mlm_train_test)))

sqrt(sum(rmse_mlm$se))

## +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

#### K-Fold Cross-Validation ####

## +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

## 10-Fold Cross-Validation
set.seed(20211118)
k10_cv <- trainControl(method = "cv", number = 10)
kf_slm <- train(egal_dem ~ gdp, data = vdem_wb, trControl = k10_cv, method = "lm")
kf_mlm <- train(egal_dem ~ ., data = vdem_wb[-1], trControl = k10_cv, method = "lm")

## Resample Comparison
kf_resample <- data.frame(kf_slm$resample, kf_mlm$resample)
kf_resample <- kf_resample[ ,!colnames(kf_resample) == "Resample.1"]
kf_resample <- kf_resample[,c(1,5,2,6,3,7,4)]
names(kf_resample)[1] = "RMSE-SLR"
names(kf_resample)[2] = "RMSE-MLR"
names(kf_resample)[3] = "R2-SLR"
names(kf_resample)[4] = "R2-MLR"
names(kf_resample)[5] = "MAE-SLR"
names(kf_resample)[6] = "MAE-MLR"

stargazer(kf_resample, type = "latex", out = "results/tables/table_01.tex", 
          summary = FALSE, rownames = FALSE, header = FALSE,style = "ajps", 
          title = "10-Fold Cross-Validation Results", 
          notes = "Source: Compiled by author using V-Dem (2020) and World Bank (2021) data.", 
          notes.align = "c")

## +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

#### LOOCV Method ####

## +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

## Leave-One-Out Cross-Validation (LOOCV)
set.seed(20211118)
loocv <- trainControl(method = "LOOCV")
loocv_slm <- train(egal_dem ~ gdp, data = vdem_wb, trControl = loocv, method = "lm")
loocv_mlm <- train(egal_dem ~ ., data = vdem_wb[-1], trControl = loocv, method = "lm")

## Comparison KF-CV and LOOCV
cv_table <- data.frame(Measure = c(names(kf_mlm$results)[2], names(kf_mlm$results)[3],
                                   names(kf_mlm$results)[4]),
                       KF_SLR = c(kf_slm$results[1,2], kf_slm$results[1,3],
                                  kf_slm$results[1,4]),
                       KF_MLR = c(kf_mlm$results[1,2], kf_mlm$results[1,3],
                                  kf_mlm$results[1,4]),
                       LOOCV_SLR = c(round(loocv_slm$results[1,2], digits = 3),
                                     round(loocv_slm$results[1,3], digits = 3),
                                     format((round(loocv_slm$results[1,4], 
                                                   digits = 3)), nsmall = 3)),
                       LOOCV_MLR = c(round(loocv_mlm$results[1,2], digits = 3),
                                     round(loocv_mlm$results[1,3], digits = 3),
                                     round(loocv_mlm$results[1,4], digits = 3)))
names(cv_table)[2] = "KF-SLR"
names(cv_table)[3] = "KF-MLR"
names(cv_table)[4] = "LOOCV-SLR"
names(cv_table)[5] = "LOOCV-MLR"

stargazer(cv_table, type = "latex", out = "results/tables/table_02.tex", 
          summary = FALSE, rownames = FALSE, header = FALSE, style = "ajps", 
          title = "10-Fold Cross-Validation and LOOCV", 
          notes = "Source: Compiled by author using V-Dem (2020) and World Bank (2021) data.", 
          notes.align = "c")

## +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

#### Shrinkage Methods ####

## +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

## Matrix of Factors and Possible Interactions
## X excluding country dummies N = 38,528
x <- model.matrix(egal_dem ~ ., data = train_set[-1])[,-1] 
x2 <- model.matrix(egal_dem ~ ., data = test_set[-1])[,-1]

## Ridge Regression
set.seed(20211118)
ridge <- cv.glmnet(x = x , y = train_set$egal_dem, family = "gaussian", 
                   alpha = 0)

## Examine Log-Lambdas to Identify the Lowest MSE
plot(ridge, sub = "Ridge Regression") 
ridge$lambda.min
log(ridge$lambda.min) ## Vertical Line 1
log(ridge$lambda.1se) ## Vertical Line 2

## MSE Ridge Model
## s = 0, OLS Model
pred_ridge <- predict(ridge, s = ridge$lambda.min, newx = x2)
mse_ridge <- mean((test_set$egal_dem - pred_ridge)^2)
coef(ridge, s = ridge$lambda.min)

## Plot Ridge Model
coef(ridge, s = "lambda.1se") %>%
  tidy() %>%
  filter(row != "(Intercept)") %>%
  top_n(70, wt = abs(value)) %>%
  ggplot(aes(value, reorder(row, value))) +
  geom_point(size = 1.5, color = "Tomato") + theme_minimal() +
  ggtitle("Influential Variables Ridge Regression") +
  xlab("Coefficient") +
  ylab("Predictors")

## LASSO Model
set.seed(20211118)
lasso <- cv.glmnet(x = x , y = train_set$egal_dem, family = "gaussian", alpha = 1)
plot(lasso, sub = "LASSO Model")
coef(lasso, s = lasso$lambda.min)

## LASSO Model
coef(lasso, s = "lambda.1se") %>%
  tidy() %>%
  filter(row != "(Intercept)") %>%
  ggplot(aes(value, reorder(row, value), color = value > 0)) +
  geom_point(show.legend = FALSE) + theme_minimal() +
  ggtitle("Influential Variables LASSO Model")  +
  xlab("Coefficient") +
  ylab("Predictors")

## MSE LASSO Model
pred_lasso <- predict(lasso, s = lasso$lambda.min, newx = x2)
mse_lasso <- mean((test_set$egal_dem - pred_lasso)^2) 

## Comparison Ridge and LASSO
mse_ridge
mse_lasso


code/stage_1_data_cleaning.R
## +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

#### Script ID ####

## +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

## Script Data Cleaning: V-Dem and World Bank
## R version 4.1.0 (2021-05-18) -- "Camp Pontanezen"
## Date: November 2021

## Bastián González-Bustamante (University of Oxford, UK)
## ORCID iD 0000-0003-1510-6820
## https://bgonzalezbustamante.com

## Regularisation and Cross-Validation: Demonstration for R
## https://github.com/bgonzalezbustamante/demo-regularisation/

## +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

#### Packages and Data ####

## +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

## Clean Environment
rm(list = ls())

## Packages
library(wbstats)
library(tidyverse)

## Data from OSF Repository
vdem <- read.csv("https://osf.io/r7v3p/download")

## Select Variables
vdem <- select(vdem, country_name, year, v2x_egaldem, v2x_corr, v2x_ex_military,
               v2x_freexp, v2x_feduni)
names(vdem)[1] = "country"
names(vdem)[3] = "egal_dem"
names(vdem)[4] = "corruption"
names(vdem)[5] = "military"
names(vdem)[6] = "free_exp"
names(vdem)[7] = "fed_uni"

## Wolrd Bank Data
new_cache <- wb_cache()
wb_set <- wb_data(indicator = c("NY.GDP.MKTP.KD.ZG", "NY.GDP.PCAP.KD.ZG", "FP.CPI.TOTL.ZG"))
## FP.CPI.TOTL.ZG: Inflation, consumer prices (annual %)
## NY.GDP.MKTP.KD.ZG: GDP growth (annual %)
## NY.GDP.PCAP.KD.ZG: GDP per capita growth (annual %)
names(wb_set)[4] = "year"
names(wb_set)[5] = "inflation"
names(wb_set)[6] = "gdp"
names(wb_set)[7] = "gdp_pc"

## Merging
vdem_wb <- left_join(vdem, wb_set[-1:-2], by = c("country", "year"))

## Save Clean Data
write.csv(vdem_wb, "data/tidy/vdem_wb.csv", fileEncoding = "UTF-8", row.names =  FALSE)

